File: Java/KokoEatingBananas.java
// O(n) from iterating through to find max in array +
// O(n log n) from doing binary search on the range, but everytime, you need to evaluate using a checker that iterates through array to find total number of hours for value k
//  -> O(n log n) 
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        // find the minimum number of bananas she has to eat per hour in order to eat all of the bananas
        
        // currently the maximum would be the largest number in the list of piles
        // the minimum possible would be 1 per hour

        // create algorithm that checks if the minimum integer k works

        // then do binary search using hte algorithm on the range 1 - max(num)
        int max = Integer.MIN_VALUE;
        for (int i = 0; i < piles.length; i++) {
            max = Math.max(max, piles[i]);
        }

        // we have our range now we can do binary search using our checker function
        int low = 1;
        int high = max;
        int mid;

        while (low <= high) {
            mid = low + (high - low) / 2;
            
            if (kChecker(piles, mid, h)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return low;


    }

    public boolean kChecker (int[] piles, int k, int h) {
        int count = 0; //count of the hours as we simulate koko eating the bananas
        for (int i = 0; i < piles.length; i++) {
            count += Math.ceil((double) piles[i] / k);
        }
        return count <= h;
    } 
    
}

File: Python/Binary Search/KokoEatingBananas.py
import math
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # how to incorporate binary search here

        # l r 

        # n piles of bananas 
        # h is the limit of hours

        # i think we can set the l to 0 and r to h
        # then have our rate checker tell us whether or not the number works

        # set l to minimum rate
        l = 1
        r = 0
        for pile in piles:
            r = max(r, pile)
        
        def rateChecker(bananaRate: int):
            hourCount = 0
            for pile in piles:
                #integer division to find num hours to eat each pile (round up)
                numHours = math.ceil(pile / bananaRate)
                hourCount += numHours
                if hourCount > h:
                    return False
            
            return hourCount <= h

        
        minHours = float("inf")
        # do our binary search

        # [3, 6, 7, 11] h = 8

        # mid = 4
        # we called rateChecker(4) -> hourCount = 1, hourCount = 3, hourCount = 7   


        while l <= r:
            mid = l + ((r - l) // 2)
            worked = rateChecker(mid)

            if worked:
                minHours = mid
                r = mid - 1
            else:
                l = mid + 1

        return minHours
